content/docs/hooks-effect.md

```diff
@@ -479,6 +479,8 @@ In the future, the second argument might get added automatically by a build-time
 >If you pass an empty array (`[]`), the props and state inside the effect will always have their initial values. While passing `[]` as the second argument is closer to the familiar `componentDidMount` and `componentWillUnmount` mental model, there are usually [better](/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies) [solutions](/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often) to avoid re-running effects too often. Also, don't forget that React defers running `useEffect` until after the browser has painted, so doing extra work is less of a problem.
 >
 >We recommend using the [`exhaustive-deps`](https://github.com/facebook/react/issues/14920) rule as part of our [`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks#installation) package. It warns when dependencies are specified incorrectly and suggests a fix.
+>
+>Also note that if you compose the dependencies array dynamically its length will not be checked, but the two arrays will be compared up to the length of the old one. When you pass variable-length arrays in second argument, always prepend the length of the array like so: `[myArray.length, ...myArray]`
```

Comment:
Thanks for the work!

I would rather note that this is not supported. React will explicitly warn: 
> Warning: The final argument passed to useEffect changed size between renders. The order and size of this array must remain constant.

The behavior about what elements are compared is not mentioned and might just be an implementation detail.

To be sure, I would propose this suggestion as an addition to the warning if this is indeed the recommended fix. The relevant warning is issued here: https://github.com/facebook/react/blob/696e736be11b22e8d4fa40438d235b63db7fddde/packages/react-reconciler/src/ReactFiberHooks.old.js#L331-L339

If we want to recommend an "always" then the warning should include it as well. If there are cases where we don't want to include `myArray.length` we should describe them here.

Reply:
I checked the comparison logic in React and it appears that it only compares up to the shorter of the two lengths so while my suggestion works the explanation is factually incorrect.

In my rather limited experience with React, hooks being dependent on the indices of a variable length array is a pretty normal phenomenon and I can't think of any other way to handle them beside what I described here, so I think an "always" is warranted.

Reply:
> I checked the comparison logic in React and it appears that it only compares up to the shorter of the two lengths so while my suggestion works the explanation is factually incorrect.

That's fine. But the source code of React does not imply a behavior we can rely on.

> so I think an "always" is warranted.

I'm not worried about whether this is warranted. If it is, then this should be part of the warning itself. So to confirm your proposal, please propose this to the React core as well and see what the core maintainers have to say. This will get a response faster.